### ISSUE TITLE
rework dependencies to make jackson+yaml+reflections optional

### ISSUE THREAD

**Issue by rmannibucau** [swagger-api/swagger-core#1632]
Hi

today swagger-jaxrs needs jackson+yaml+reflections but actually all of them can be optional:
- jackson can be replaced by container json provider (johnzon for instance works fine with swagger-ui)
- yaml is optional since you can use only json features
- reflections can be replaced by DefaultJaxrsScanner or another scanner implementation (so BeanConfig shouldn't force it)

So code would need to be reworked to not force to use such dependencies when not actually needed at runtime (see https://github.com/rmannibucau/swagger-in-tomee for instance where these dependencies are there only cause otherwise swagger doesn't start).

**Comment by webron** [comment]
We literally depend on Jackson's processing, so we cannot depend on something else. YAML, being a first class citizen of the OpenAPI spec, will remain an output of the library and cannot become optional.

**Comment by rmannibucau** [comment]
Not sure I follow you:

- yaml support as a feature IS optional. Typically just being interested in swagger-ui makes it optional since default is json.
- jackson is not needed for serialization and a spi can avoid it.

My point is mainly than a .war of 100k becomes a fat war of 5M with swagger stack and most of it is useless at runtime for several cases if swagger is a bit refactored

**Comment by webron** [comment]
YAML support is as optional as JSON support is.

As for Jackson - check out https://github.com/swagger-api/swagger-core/tree/master/modules/swagger-core/src/main/java/io/swagger/jackson. Our code depends on Jackson.

However, if you feel strongly about it and think it can be done, by all means, feel free to submit a PR. If it makes sense and is usable (*with* YAML support), then it could be great for everyone.

**Comment by rmannibucau** [comment]
Ok missed the fact jackson is also used as a parser. Do you have any idea if doing a minimized - as the option of the maven plugin - shade would help?

Going down to 1 or 2M max would be already awesome.

**Comment by webron** [comment]
I have no idea, but I was serious about the PR. If you can make it smaller in a way that doesn't harm the functionality of the project, doesn't break backward compatibility and would benefit the users of the project, then by all means, it would be great.

**Comment by rmannibucau** [comment]
Ok, to move a bit forward on this topic here is a list of dep sizes (removed the not really needed ones since they should be provided like javax.* and slf4j):

```
2,2M guava-18.0.jar
1,1M jackson-databind-2.4.5.jar
732K javassist-3.19.0-GA.jar
561K joda-time-2.2.jar
377K commons-lang3-3.2.1.jar
314K jackson-dataformat-yaml-2.4.5.jar
265K snakeyaml-1.12.jar
221K jackson-core-2.4.5.jar
159K stax2-api-3.1.4.jar
133K swagger-models-1.5.7-SNAPSHOT.jar
127K reflections-0.9.10.jar
103K servlet-api-2.5.jar
101K swagger-core-1.5.7-SNAPSHOT.jar
 89K jackson-dataformat-xml-2.4.5.jar
 73K annotations-2.0.1.jar
 55K jackson-datatype-joda-2.4.5.jar
 38K jackson-annotations-2.4.5.jar
 32K jackson-module-jaxb-annotations-2.4.5.jar
 28K jackson-jaxrs-base-2.4.5.jar
 17K jackson-jaxrs-json-provider-2.4.5.jar
 16K swagger-annotations-1.5.7-SNAPSHOT.jar
```

So removing guava would already be a super huge win (-2.2M). Check quickly how it is used and it looks possible even if not a 1h work.
Then another easy win is to remove from io.swagger.jaxrs.config.BeanConfig reflections (maybe creating a JaxRsBeanConfig which relies on DefaultJaxRsScanner in parallel) which would remove the need of reflections+javassist (-1M for plain jaxrs or CXF users).

With these modifications we already gained 3M so this issue can be considered as done. Going further means working on jackson part which as you mentionned can be more hardcore since it means replacing the model parsing by custom code.

Once done a test with a minimized shade can be done to check if it works and gain anything but we would already have moved a lot from today state.

wdyt?

**Comment by davsclaus** [comment]
+1

Yeah guave is a pain they break the apis, and you end up with X use guave 15 and another 17 and then now 18 and they cant work together.

**Comment by ilgrosso** [comment]
+1 from me as well, and I am available to update the CXF Swagger counterparts when done

**Comment by webron** [comment]
We shouldn't be looking to reduce dependencies at the expense of adding complexity to the code, that's my take. On the other hand, removing dependencies will ease up integration and dependency conflicts. Personally, I'm less concerned with the size of the dependency and until now nobody even mentioned it (ever since we moved to pure Java code).

If you'd like to approach this, I'd suggest tackling it one by one (that is multiple isolated PRs and not a huge change in one). I cannot guarantee this will be merged without seeing the changes first, obviously, but that would give us a better idea of your approach.

I'd leave out Jackson for now.

Thank you for taking the time to even investigate it - users like you really help us push the project forward.

**Comment by webron** [comment]
To clarify, I meant leave Jackson out of the process. Don't try to remove it as a dependency.

**Comment by fehguy** [comment]
Here is a suggestion on this topic.  While it would be nice if *everything* was pluggable, I don't know if that would serve the community well.  If we really don't want runtime introspection of classes (i.e. automatically generating the swagger definition), it's quite simply to build a static swagger.json, stuff it in your webapp folder and call it done. Then you have _zero_ dependencies on the stack.

If someone wants to take a stab at helping make things pluggable, yet require minimal configuration (like it has), I'd love to see that.

I do believe we support the most general case (5mb of dependencies is honestly nothing if you step back a bit) BUT making it more flexible--if testable--would never hurt.

**Comment by rmannibucau** [comment]
Well with guava PR and the PR allowing to integration without javassist (reflections) or using WebXMLReader with jaxrs module (doesnt need reflections stack) it is 2/3M which is no more a blocker.

Side note: 5M is a lot most of the time. Today with JavaEE servers we often deploy skinny wars.

So only missing thing to this issue is reflection part which I agree is not very easy to replace - understand we can ignore it for this issue.

**Comment by fehguy** [comment]
Yes you understand though that those PRs are not passing tests and cannot be merged?

**Comment by rmannibucau** [comment]
Guava one is for sure, didnt check others but I guess WebXmlReader one will surely need some review anyway cause it changes the config a bit
